### PR TITLE
TRUNK-3624: Created Program data type support

### PR DESCRIPTION
**Pull request by foolchan2556**

### CODE REVIEW

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+package org.openmrs.customdatatype.datatype;
+
+import org.openmrs.Program;
+import org.openmrs.customdatatype.SerializingCustomDatatype;
+
+import com.thoughtworks.xstream.XStream;
+import com.thoughtworks.xstream.io.xml.DomDriver;
+
+public class ProgramDatatype extends SerializingCustomDatatype<Program> {
```
Comment (rkorytkowski) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java]: Note that other datatypes are annotated with @Component so that they can be autowired by DatatypeService. Please put @Component here and say in javadoc @since 1.12

File: web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import org.openmrs.customdatatype.CustomDatatype;
+import org.openmrs.customdatatype.CustomDatatype.Summary;
+import org.openmrs.customdatatype.InvalidCustomValueException;
+import org.openmrs.customdatatype.datatype.ProgramDatatype;
+import org.openmrs.serialization.SerializationException;
+import org.openmrs.serialization.SimpleXStreamSerializer;
+import org.springframework.stereotype.Component;
+
+/**
+ * Handler for the Program custom datatype
```
Comment (rkorytkowski) [inline web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java]: Please add @since 1.12

File: web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		catch (SerializationException ex) {
+			throw new InvalidCustomValueException("Invalid Program " + fieldValue);
+		}
+	}
+	
+	/**
+	 * @see org.openmrs.web.attribute.handler.HtmlDisplayableDatatypeHandler#toHtmlSummary(org.openmrs.customdatatype.CustomDatatype, java.lang.String)
+	 */
+	@Override
+	public Summary toHtmlSummary(CustomDatatype<Program> datatype, String valueReference) {
```
Comment (rkorytkowski) [inline web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java]: Have you read toHtmlSummary javadoc before implementing it? The summary should be shorter than 100 characters... It could be Program.name for example.

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.customdatatype.datatype;
+
+import org.openmrs.Program;
+import org.openmrs.customdatatype.SerializingCustomDatatype;
+
+import com.thoughtworks.xstream.XStream;
+import com.thoughtworks.xstream.io.xml.DomDriver;
+
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java]: The ticket description was updated, please extend BaseMetadatatype

File: web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import org.openmrs.customdatatype.InvalidCustomValueException;
+import org.openmrs.customdatatype.datatype.ProgramDatatype;
+import org.openmrs.serialization.SerializationException;
+import org.openmrs.serialization.SimpleXStreamSerializer;
+import org.springframework.stereotype.Component;
+
+/**
+ * Handler for the Program custom datatype
+ */
+@Component
```
Comment (wluyima) [inline web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java]: Extend BaseMetadataFieldGenDatatypeHandler

Comment (foolchan2556) [inline web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java]: Thanks man! I will take care of that. BTW, do you know which part of the Program data shall I serialize as xml?

Comment (djazayeri) [inline web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java]: You don't need to "serialize" exactly. Just write out Program.uuid (and
read it back).

Comment (foolchan2556) [inline web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java]: I see, thank you

Comment (wluyima) [inline web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java]: You won't need this actually once you extend BaseOpenmrsDatatype

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java [hunk 6]
```diff
@@ -9,23 +9,16 @@
...
-	
-	/**
-	 * @see BaseOpenmrsDatatype#doGetTextSummary(Object)
-	 * @should use the name in summary instance
-	 */
+public class ProgramDatatype<T extends OpenmrsObject> extends BaseOpenmrsDatatype<T> {
 	@Override
-	public Summary doGetTextSummary(T typedValue) {
-		return new CustomDatatype.Summary(typedValue.getName(), true);
+	public T deserialize(String serializedValue){
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java]: Please can you undo this change, the original implementation is the correct one

File: web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+import org.openmrs.serialization.SimpleXStreamSerializer;
+import org.springframework.stereotype.Component;
+
+/**
+ * Handler for the Program custom datatype
+ *
+ * @Component
+ * @since 1.12
+ */
+public class ProgramFieldGenDatatypeHandler<DT extends BaseMetadataDatatype<T>, T extends OpenmrsMetadata> extends BaseMetadataFieldGenDatatypeHandler<DT, T> {
```
Comment (wluyima) [inline web/src/main/java/org/openmrs/web/attribute/handler/ProgramFieldGenDatatypeHandler.java]: The first type parameter value should be **ProgramDatatype** and not **DT extends BaseMetadataDatatype<T>**

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java [hunk 8]
```diff
@@ -9,23 +9,16 @@
...
  * 
  * @since 1.12
  */
-public abstract class BaseMetadataDatatype<T extends OpenmrsMetadata> extends BaseOpenmrsDatatype<T> {
-	
-	/**
-	 * @see BaseOpenmrsDatatype#doGetTextSummary(Object)
-	 * @should use the name in summary instance
-	 */
+public class ProgramDatatype<T extends OpenmrsObject> extends BaseOpenmrsDatatype<T> {
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java]: The type parameter value should be **Program**

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java (hunk 6 above)
Comment (foolchan2556) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java]: @wluyima Hey, thanks! But deserialize must be overrided or I have to declare ProgramDatatype as an abstract class.

Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java]: Sorry, yes you need to implement deserialize, and the typical implementation is to fetch the object by uuid from the appropriate service since serialize from BaseOpenmrsDatatype does return the uuid.

File: api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java (hunk 8 above)
Comment (foolchan2556) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java]: Thanks！ What would you suggest as an appropriate service? Since I really know little abt this part actually. Just in case you know that, or I can try to figure it out myself.

Comment (wluyima) [inline api/src/main/java/org/openmrs/customdatatype/datatype/ProgramDatatype.java]: ProgramWorkflowService

**Comment by foolchan2556** [comment]
Have cleaned commit history and opened another PR
